src/celeritas/phys/PhysicsParams.cc

```diff
@@ -584,6 +591,145 @@ void PhysicsParams::build_xs(const Options&        opts,
     }
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Construct model cross section CDFs.
+ */
+void PhysicsParams::build_model_xs(const MaterialParams& mats,
+                                   HostValue*            data) const
+{
+    CELER_EXPECT(*data);
+
+    ValueGridInserter insert_grid(&data->reals, &data->value_grids);
+
+    // Micro xs grid IDs for each model and applicable particle, each material,
+    // and each element in the material
+    std::vector<std::vector<std::vector<ValueGridId>>> temp_grid_ids(
+        data->model_ids.size());
+    size_type pm_idx{0};
+
+    for (auto model_idx : range(this->num_models()))
+    {
+        const Model& model = *models_[model_idx].first;
+
+        // Loop over applicable particles
+        for (Applicability applic : model.applicability())
+        {
+            for (auto mat_id : range(MaterialId{mats.size()}))
+            {
+                applic.material = mat_id;
+                auto material   = mats.get(mat_id);
+
+                // TODO: Create combined SB + RB micro xs grids or possibly
+                // remove combined bremsstrahlung model
+                CELER_VALIDATE(!(dynamic_cast<const CombinedBremModel*>(&model)
+                                 && material.num_elements() > 1),
+                               << "model '" << model.label()
+                               << "' cannot be used with materials composed "
+                                  "of more than one element (material '"
+                               << mats.id_to_label(mat_id) << "' has "
+                               << material.num_elements() << " elements)");
+
+                // Construct microscopic cross section builders
+                auto builders = model.micro_xs(applic);
+                if (builders.empty())
+                {
+                    // Models that calculate xs on the fly and models
+                    // with material-independent discrete interactions
+                    // won't have micro xs grids
+                    continue;
+                }
+                CELER_ASSERT(builders.size() == material.num_elements());
+
+                // Construct grids for each element in the material
+                CELER_ASSERT(pm_idx < temp_grid_ids.size());
+                temp_grid_ids[pm_idx].resize(mats.size());
```

Comment:
This will only have an effect the first iteration out of all the materials, but I don't see a better way of doing this 🤔 unless we add flags to the model to denote on-the-fly calculation, etc...

Definitely leave this as is for now.